File: graph/src/100_same_tree.cpp
#include <iostream>

struct TreeNode{
    int val;
    struct TreeNode* left;
    struct TreeNode* right;
};

bool isSameTree(TreeNode* p, TreeNode* q) {
    if(p == NULL && q == NULL){
        return true;
    }else{
        if(p == NULL || q == NULL){
            return false;
        }
        if(p->val == q->val){
            return isSameTree(p->left, q->left)&&isSameTree(p->right,q->right);
        }else{
            return false;
        }
    }
}

int main()
{
    struct TreeNode a = {1};
    struct TreeNode b = {2};
    printf("two tree nodes are tree: %d\n",isSameTree(&a,&b));
    
    struct TreeNode c = {1};
    struct TreeNode d = {2};
    
    a.left = &d;
    c.left = &b;
    std::cout<<"two tree nodes are true? "<<isSameTree(&a,&c)<<std::endl;
    return 0;
}

File: graph/src/100_same_tree.py
from typing import Optional, TypeVar

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def isSameTree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    if p == None and q == None:
        return True
    else:
        if p == None or q == None:
            return False
        else:
            return (p.val == q.val) and isSameTree(p.left,q.left) and isSameTree(p.right,q.right)

if __name__ == '__main__':
    a = TreeNode(1)
    b = TreeNode(2)
    c = TreeNode(3)
    d = TreeNode(1)

    print(f'a is same as b: {isSameTree(a,b)}')
    
    a.left = b
    d.left = b
    print(f'a is same as d: {isSameTree(a,d)}')

    a.right = c
    d.right = b
    print(f'a is same as d: {isSameTree(a,d)}')
